File: compat/suites/node-js-sdk/src/groups/cognito.ts
/**
 * groups/cognito.ts — Cognito User Pools compatibility test groups for the Node.js suite.
 *
 * Status: NOT implemented in Overcast. All tests expected to fail with 501.
 * These tests define the coverage target for future Cognito implementation.
 *
 * Groups:
 *   cognito-token-validity — user pool client token validity settings
 *
 * cognito-userpools resolves through its authored scenario
 * (compat/model/authored/cognito-userpools.json).
 */

import {
  CreateUserPoolCommand,
  DeleteUserPoolCommand,
  CreateUserPoolClientCommand,
  DeleteUserPoolClientCommand,
  DescribeUserPoolClientCommand,
  UpdateUserPoolClientCommand,
} from "@aws-sdk/client-cognito-identity-provider";
import { makeClients } from "../lib/clients.ts";
import type { TestGroup } from "../lib/harness.ts";
import * as assert from "node:assert/strict";

export function makeCognitoGroups(suite: string): TestGroup[] {
  return [
    // ── cognito-token-validity ─────────────────────────────────────────────
    {
      suite,
      service: "cognito",
      name: "cognito-token-validity",
      tests: [
        {
          name: "CreateUserPoolClientWithTokenValidity",
          fn: async (ctx) => {
            const { cognito } = makeClients(ctx);
            const poolName = `compat-tv-${ctx.runId}`;
            const poolResp = await cognito.send(
              new CreateUserPoolCommand({ PoolName: poolName }),
            );
            const poolId = poolResp.UserPool?.Id;
            assert.ok(poolId, "CreateUserPool: missing Id");
            (ctx as Record<string, unknown>)["_tvPoolId"] = poolId;

            const resp = await cognito.send(
              new CreateUserPoolClientCommand({
                UserPoolId: poolId,
                ClientName: `compat-client-${ctx.runId}`,
                AccessTokenValidity: 2,
                IdTokenValidity: 3,
                RefreshTokenValidity: 7,
                TokenValidityUnits: {
                  AccessToken: "hours",
                  IdToken: "hours",
                  RefreshToken: "days",
                },
              }),
            );
            const client = resp.UserPoolClient;
            assert.ok(
              client?.ClientId,
              "CreateUserPoolClient: missing ClientId",
            );
            assert.equal(client.AccessTokenValidity, 2);
            assert.equal(client.IdTokenValidity, 3);
            assert.equal(client.RefreshTokenValidity, 7);
            assert.equal(client.TokenValidityUnits?.AccessToken, "hours");
            assert.equal(client.TokenValidityUnits?.IdToken, "hours");
            assert.equal(client.TokenValidityUnits?.RefreshToken, "days");
            (ctx as Record<string, unknown>)["_tvClientId"] = client.ClientId;
          },
        },
        {
          name: "DescribeUserPoolClientTokenValidity",
          fn: async (ctx) => {
            const { cognito } = makeClients(ctx);
            const poolId = (ctx as Record<string, unknown>)[
              "_tvPoolId"
            ] as string;
            const clientId = (ctx as Record<string, unknown>)[
              "_tvClientId"
            ] as string;
            assert.ok(poolId && clientId, "no pool/client from create");

            const resp = await cognito.send(
              new DescribeUserPoolClientCommand({
                UserPoolId: poolId,
                ClientId: clientId,
              }),
            );
            const client = resp.UserPoolClient;
            assert.equal(client?.AccessTokenValidity, 2);
            assert.equal(client?.IdTokenValidity, 3);
            assert.equal(client?.RefreshTokenValidity, 7);
          },
        },
        {
          name: "UpdateUserPoolClientTokenValidity",
          fn: async (ctx) => {
            const { cognito } = makeClients(ctx);
            const poolId = (ctx as Record<string, unknown>)[
              "_tvPoolId"
            ] as string;
            const clientId = (ctx as Record<string, unknown>)[
              "_tvClientId"
            ] as string;
            assert.ok(poolId && clientId, "no pool/client from create");

            const resp = await cognito.send(
              new UpdateUserPoolClientCommand({
                UserPoolId: poolId,
                ClientId: clientId,
                AccessTokenValidity: 30,
                TokenValidityUnits: {
                  AccessToken: "minutes",
                  IdToken: "hours",
                  RefreshToken: "days",
                },
              }),
            );
            const client = resp.UserPoolClient;
            assert.equal(client?.AccessTokenValidity, 30);
            assert.equal(client?.TokenValidityUnits?.AccessToken, "minutes");
          },
        },
        {
          name: "DeleteUserPoolClient",
          fn: async (ctx) => {
            const { cognito } = makeClients(ctx);
            const poolId = (ctx as Record<string, unknown>)[
              "_tvPoolId"
            ] as string;
            const clientId = (ctx as Record<string, unknown>)[
              "_tvClientId"
            ] as string;
            if (!poolId || !clientId) return;
            await cognito.send(
              new DeleteUserPoolClientCommand({
                UserPoolId: poolId,
                ClientId: clientId,
              }),
            );
          },
        },
      ],
      teardown: async (ctx) => {
        const { cognito } = makeClients(ctx);
        const poolId = (ctx as Record<string, unknown>)["_tvPoolId"] as string;
        if (!poolId) return;
        const clientId = (ctx as Record<string, unknown>)[
          "_tvClientId"
        ] as string;
        if (clientId) {
          try {
            await cognito.send(
              new DeleteUserPoolClientCommand({
                UserPoolId: poolId,
                ClientId: clientId,
              }),
            );
          } catch {}
        }
        try {
          await cognito.send(new DeleteUserPoolCommand({ UserPoolId: poolId }));
        } catch {}
      },
    },
  ];
}

File: compat/suites/python-sdk/groups/cognito.py
"""
groups/cognito.py — Cognito User Pools compatibility test implementations for the Python suite.

Groups: cognito-token-validity. cognito-userpools resolves through its
authored scenario (compat/model/authored/cognito-userpools.json).
"""

from __future__ import annotations
from lib.harness import TestContext
from lib.clients import make_clients


def _cognito(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region)._get("cognito-idp")


# ── ImplMap ───────────────────────────────────────────────────────────────────

IMPLS = {
    "cognito-token-validity:CreateUserPoolClientWithTokenValidity": lambda ctx: CreateClientTokenValidity(ctx),
    "cognito-token-validity:DescribeUserPoolClientTokenValidity": lambda ctx: DescribeClientTokenValidity(ctx),
    "cognito-token-validity:UpdateUserPoolClientTokenValidity": lambda ctx: UpdateClientTokenValidity(ctx),
    "cognito-token-validity:DeleteUserPoolClient": lambda ctx: DeleteUserPoolClientFn(ctx),
}

SETUP = {
    "cognito-token-validity": lambda ctx: None,
}
TEARDOWN = {
    "cognito-token-validity": lambda ctx: _teardown_token_validity(ctx),
}


# ── cognito-token-validity ────────────────────────────────────────────────────

def _teardown_token_validity(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_id = ctx.get("tv_pool_id")
    client_id = ctx.get("tv_client_id")
    if pool_id and client_id:
        try:
            cog.delete_user_pool_client(UserPoolId=pool_id, ClientId=client_id)
        except Exception:
            pass
    if pool_id:
        try:
            cog.delete_user_pool(UserPoolId=pool_id)
        except Exception:
            pass


def CreateClientTokenValidity(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_name = f"compat-tv-{ctx.run_id}"
    resp = cog.create_user_pool(PoolName=pool_name)
    pool_id = resp.get("UserPool", {}).get("Id")
    if not pool_id:
        raise AssertionError("CreateClientTokenValidity: missing pool Id")
    ctx["tv_pool_id"] = pool_id

    resp = cog.create_user_pool_client(
        UserPoolId=pool_id,
        ClientName=f"compat-client-{ctx.run_id}",
        AccessTokenValidity=2,
        IdTokenValidity=3,
        RefreshTokenValidity=7,
        TokenValidityUnits={
            "AccessToken": "hours",
            "IdToken": "hours",
            "RefreshToken": "days",
        },
    )
    client = resp.get("UserPoolClient", {})
    if not client.get("ClientId"):
        raise AssertionError("CreateClientTokenValidity: missing ClientId")
    ctx["tv_client_id"] = client["ClientId"]


def DescribeClientTokenValidity(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_id = ctx.get("tv_pool_id")
    client_id = ctx.get("tv_client_id")
    if not pool_id or not client_id:
        raise AssertionError("DescribeClientTokenValidity: missing pool/client")
    cog.describe_user_pool_client(UserPoolId=pool_id, ClientId=client_id)


def UpdateClientTokenValidity(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_id = ctx.get("tv_pool_id")
    client_id = ctx.get("tv_client_id")
    if not pool_id or not client_id:
        raise AssertionError("UpdateClientTokenValidity: missing pool/client")
    cog.update_user_pool_client(
        UserPoolId=pool_id,
        ClientId=client_id,
        AccessTokenValidity=30,
        TokenValidityUnits={
            "AccessToken": "minutes",
            "IdToken": "hours",
            "RefreshToken": "days",
        },
    )


def DeleteUserPoolClientFn(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_id = ctx.get("tv_pool_id")
    client_id = ctx.get("tv_client_id")
    if not pool_id or not client_id:
        return
    cog.delete_user_pool_client(UserPoolId=pool_id, ClientId=client_id)
